documentation/Architecture/The-Data-Processing-Pipeline.md

```diff
@@ -1,43 +1,131 @@
 
-Using the Data Processor, we can efficiently access information from multiple binaries and reuse data.
-Data can be easily shared and the developer is granted maximal flexibility for modular code.
+# Pipeline Components
 
-A Data Processor contains Source Parsers and a Data Cookers.
+At the highest level, when a plugin wishes to create a data processing pipeline, it will define two types of 
+components: __source parsers__ and __data cookers__.
 
-![](.attachments/DataProcessingPipeline.png)
+## Source Parsers
 
+A `SourceParser` is a class whose sole purpose is to process (parse) the raw data inside the data sources a 
+`CustomDataProcessor` must process. Conceptually, a `CustomDataProcessor` further *delegates* the task of 
+processing its data sources to a `SourceParser`.
 
-# Source Parser
+> :question: This delegation raises the question: if a data processor can just further delegate work, 
+> why have the data processor in the first place? The reason for having both classes is that
+> 1) `CustomDataProcessor`s were introduced well before `SourceParser`s and
+> 2) Having both provides developers with more flexibility on how they structure their plugins
 
-The Source Parser is bound to a procesor which is created by a CDS in order to:
-* read in raw data from the Data Source 
-* create a standardized data format
+From an interface perspective, a `SourceParser` has the following properties:
+* A `SourceParser` has a `string` identifier
+* A `SourceParser` is bound to a specific `CustomDataProcesor`
+* A `SourceParser` must, at some point, process the data sources originally given to the `CustomDataSource` which  
+  created the `CustomDataProcessor` the parser is bound to 
+* A `SourceParser` will, *while processing its data sources*, emit each individual data point contained 
+  within the data sources it is processing
 
-The standardized data is then sent to a data cooker.
+The data points a `SourceParser` emits are called __events__, and each __event__ must have a *key*. Conceptually, 
+the emitted events are `(key, value)` pairs contained within a single class, where each value is a data point and its 
+key is some property of the data point. 
 
-# Data Cooker
+A `SourceParser` emits each event one at a time. __It is highly encouraged to make sure a source parser 
+only parses/iterates the data inside its data sources once, emitting each event as soon as it is 
+encountered.__
 
-The Data Cooker consumes events sent from the Source Parser and specifies the data keys. 
-This allows for the standardized data to be read into a C# data structure and allows programmatic access to any data.
+## Data Cookers
 
-The platform Data Cookers and Tables are built on is extensible by default. This means Data Cookers can not only be referenced by 
-eachother but also referenced in other binaries and other projects.
+A `DataCooker` is a class whose purpose is to implement the data transformations outlined at the beginning of this document. 
```

Comment:
this should be edited to this to keep with the same format as before but also to keep the document readable. You don't want the reader to have to jump up and around the document multiple time -- "A DataCooker is a class with the main objective of  data transformations to allow (1) transformation of data to any data types, (2) [I don't understand this one well enough to reword it but maybe -- allow external uses outside the CDP], (3) From an interface perspective, a DataCooker has the following properties:"

Reply:
Changed this to:

A `DataCooker` is a class that is responsible for implementing the data transformations whose 
output is accessible by both other plugin modules and external binaries. 
From an interface perspective...

How does this sound?

Reply:
its more descriptive. works great!